src/main/kotlin/cacophony/regex/RegexUtil.kt

```diff
@@ -0,0 +1,81 @@
+package cacophony.regex
+
+import cacophony.utils.AlgebraicRegex
+import cacophony.utils.AlgebraicRegex.AtomicRegex
+import cacophony.utils.AlgebraicRegex.ConcatenationRegex
+import cacophony.utils.AlgebraicRegex.StarRegex
+import cacophony.utils.AlgebraicRegex.UnionRegex
+
+class RegexSyntaxErrorException : Exception()
+
+fun algebraicRegexEquals(
+    x: AlgebraicRegex,
+    y: AlgebraicRegex,
+): Boolean {
+    when (x) {
+        is AtomicRegex ->
+            if (y is AtomicRegex) {
+                return x.symbol == y.symbol
+            }
+        is ConcatenationRegex ->
+            if (y is ConcatenationRegex) {
+                return x.internalRegexes.size == y.internalRegexes.size &&
+                    x.internalRegexes.zip(y.internalRegexes).all {
+                            (a, b) ->
+                        algebraicRegexEquals(a, b)
+                    }
+            }
+        is StarRegex ->
+            if (y is StarRegex) {
+                return algebraicRegexEquals(x.internalRegex, y.internalRegex)
+            }
+        is UnionRegex ->
+            if (y is UnionRegex) {
+                return x.internalRegexes.size == y.internalRegexes.size &&
+                    x.internalRegexes.zip(y.internalRegexes).all {
+                            (a, b) ->
+                        algebraicRegexEquals(a, b)
```

Comment:
Does it cover for example x=(a|b), y=(b|a)?

Reply:
Only for tests, it also does not work for something like (a|b|c) and (a|(b|c))

Reply:
Moved to test directory